File: Week3/suitescript-je-review/src/FileCabinet/SuiteScripts/je_review/je_rules.js
/**
 * je_rules.js  -  THE DETERMINISTIC RULE ENGINE (ported 1:1 from Week3/je_rules.py).
 *
 * Ten pure functions raise the flags; a deterministic score maps them to a risk tier. No AI, no N
 * modules -> unit-testable off-platform (see __tests__/je_rules.test.js). The N/llm layer only comes
 * afterwards to judge the grey-zone cases and challenge itself; it never decides a flag or a figure.
 *
 * An entry (assembled by the orchestrator from je_sql results) has:
 *   je_id, entry_date(Date), post_ts(Date), period_id, account, account_type(normalised category),
 *   amount(Number), dr_cr, preparer, approver, description, has_support(bool).
 *
 * @NApiVersion 2.1
 */
define([], function () {

  var SEVERITY_WEIGHT = { high: 3, medium: 2, low: 1 };

  var GENERIC_DESC = {
    adjustment: 1, adj: 1, reclass: 1, correction: 1, 'to correct': 1, 'true up': 1,
    'true-up': 1, plug: 1, misc: 1, journal: 1, entry: 1
  };

  /**
   * Map a raw NetSuite accttype to the category the rules use. Control = AP/AR subledger accounts;
   * bank = bank/clearing; equity = equity/retained earnings. Everything else is expense/revenue/other.
   */
  function normalizeAcctType(accttype) {
    switch (String(accttype || '')) {
      case 'Bank': return 'bank';
      case 'AcctPay': case 'AcctRec': return 'control';
      case 'Equity': case 'RetEarnings': return 'equity';
      case 'Income': case 'OthIncome': return 'revenue';
      case 'Expense': case 'COGS': case 'OthExpense': return 'expense';
      default: return 'other';
    }
  }

  function flag(rule, severity, detail) { return { rule: rule, severity: severity, detail: detail }; }

  function money(v) { return (Math.round(Number(v) || 0)).toLocaleString('en-US'); }

  // days between two Dates (b - a), calendar days.
  function dayDiff(a, b) {
    var ms = 24 * 60 * 60 * 1000;
    return Math.round((toDate(b).getTime() - toDate(a).getTime()) / ms);
  }
  function toDate(d) { return (d instanceof Date) ? d : new Date(d); }

  // ==========================================================================
  // The rules — each pure: (entry, ctx) -> flag | null
  // ==========================================================================
  function ruleOverThreshold(je, ctx) {
    if (Number(je.amount) >= ctx.approvalThreshold) {
      return flag('over_threshold', 'medium',
        'amount ' + money(je.amount) + ' >= approval threshold ' + money(ctx.approvalThreshold));
    }
    return null;
  }

  function ruleRoundThousand(je, ctx) {
    var amt = Number(je.amount);
    if (amt >= ctx.roundMin && amt % 1000 === 0) {
      return flag('round_thousand', 'low', 'amount ' + money(amt) + ' is an exact multiple of 1,000');
    }
    return null;
  }

  function ruleOffHours(je, ctx) {
    var ts = toDate(je.post_ts);
    var reasons = [];
    var dow = ts.getDay();                       // 0=Sun .. 6=Sat
    if (dow === 0 || dow === 6) reasons.push('weekend');
    if (isHoliday(ts, ctx.holidays)) reasons.push('holiday');
    var h = ts.getHours();
    if (h < ctx.businessStart || h >= ctx.businessEnd) {
      reasons.push(pad(h) + ':' + pad(ts.getMinutes()) + ' outside ' +
        pad(ctx.businessStart) + ':00-' + pad(ctx.businessEnd) + ':00');
    }
    if (reasons.length) {
      return flag('off_hours', 'low', 'posted ' + reasons.join(', ') + ' (' + fmtTs(ts) + ')');
    }
    return null;
  }

  function ruleWeakDescription(je, ctx) {
    var desc = String(je.description || '').trim();
    if (!desc) return flag('weak_description', 'medium', 'description is blank');
    var low = desc.toLowerCase();
    if (desc.length < ctx.shortDescLen || GENERIC_DESC.hasOwnProperty(low)) {
      return flag('weak_description', 'low', 'description uninformative: "' + desc + '"');
    }
    return null;
  }

  function ruleNoSupport(je, ctx) {
    if (!je.has_support) {
      var sev = Number(je.amount) >= ctx.approvalThreshold ? 'high' : 'medium';
      return flag('no_support', sev, 'no supporting document (amount ' + money(je.amount) + ')');
    }
    return null;
  }

  function ruleSensitiveAccount(je, ctx) {
    if (ctx.sensitiveTypes[je.account_type]) {
      return flag('sensitive_account', 'high',
        'manual JE to ' + je.account_type + ' account "' + je.account + '"');
    }
    return null;
  }

  function ruleEntryPostGap(je, ctx) {
    var gap = dayDiff(je.entry_date, je.post_ts);
    if (gap >= ctx.gapDays) {
      return flag('entry_post_gap', 'medium',
        gap + ' days between entry date ' + fmtDate(je.entry_date) + ' and posting ' + fmtDate(je.post_ts));
    }
    return null;
  }

  function ruleClosedPeriod(je, ctx) {
    if (ctx.closedPeriodIds[String(je.period_id)]) {
      return flag('closed_period', 'high', 'posted into closed period ' + (je.period || je.period_id));
    }
    return null;
  }

  function ruleSodBreach(je, ctx) {
    var preparer = String(je.preparer || '').trim();
    var approver = String(je.approver || '').trim();
    if (approver && preparer && preparer === approver) {
      return flag('sod_breach', 'high', 'preparer and approver are the same person (' + preparer + ')');
    }
    if (!approver) {
      return flag('sod_breach', 'medium', 'no independent approver recorded (preparer ' + (preparer || 'n/a') + ')');
    }
    return null;
  }

  function ruleNearDuplicate(je, ctx) {
    var reg = ctx.register || [];
    for (var i = 0; i < reg.length; i++) {
      var other = reg[i];
      if (other === je || other.je_id === je.je_id) continue;
      if (other.account === je.account &&
          Math.abs(Number(other.amount) - Number(je.amount)) < 0.01 &&
          Math.abs(dayDiff(je.post_ts, other.post_ts)) <= ctx.dupWindowDays) {
        return flag('near_duplicate', 'medium',
          'matches ' + other.je_id + ': same account "' + je.account + '", amount ' +
          money(je.amount) + ', within ' + ctx.dupWindowDays + 'd');
      }
    }
    return null;
  }

  var RULES = [
    ruleOverThreshold, ruleRoundThousand, ruleOffHours, ruleWeakDescription, ruleNoSupport,
    ruleSensitiveAccount, ruleEntryPostGap, ruleClosedPeriod, ruleSodBreach, ruleNearDuplicate
  ];

  // ==========================================================================
  // Evaluation + deterministic scoring (mirrors the Python score()/assess())
  // ==========================================================================
  // Rules that depend on an approval workflow; skipped when ctx.enableApproverRules is false.
  var APPROVER_RULES = [ruleSodBreach];

  function evaluate(je, ctx) {
    var flags = [];
    for (var i = 0; i < RULES.length; i++) {
      if (ctx.enableApproverRules === false && APPROVER_RULES.indexOf(RULES[i]) !== -1) continue;
      var f = RULES[i](je, ctx);
      if (f) flags.push(f);
    }
    return flags;
  }

  function score(flags) {
    if (!flags.length) return { risk_score: 0, tier: 'clear' };
    var total = 0, hasHigh = false;
    flags.forEach(function (f) { total += SEVERITY_WEIGHT[f.severity]; if (f.severity === 'high') hasHigh = true; });
    var tier = (hasHigh || total >= 5) ? 'high' : (total >= 2 ? 'medium' : 'low');
    return { risk_score: total, tier: tier };
  }

  function assess(je, ctx) {
    var flags = evaluate(je, ctx);
    var s = score(flags);
    return { je: je, flags: flags, risk_score: s.risk_score, tier: s.tier,
             flag_rules: flags.map(function (f) { return f.rule; }) };
  }

  /** Assess a whole register. Points ctx.register at it so near-duplicate sees every other row. */
  function assessRegister(register, ctx) {
    ctx.register = register;
    return register.map(function (je) { return assess(je, ctx); });
  }

  // Build a RuleContext from a config object (defaults mirror je_rules.RuleContext in Python).
  function contextFromConfig(cfg, closedPeriodIds, holidays) {
    return {
      register: [],
      closedPeriodIds: closedPeriodIds || {},
      holidays: holidays || [],
      approvalThreshold: cfg.approvalThreshold != null ? cfg.approvalThreshold : 50000,
      roundMin: cfg.roundMin != null ? cfg.roundMin : 10000,
      gapDays: cfg.gapDays != null ? cfg.gapDays : 30,
      businessStart: cfg.businessStart != null ? cfg.businessStart : 7,
      businessEnd: cfg.businessEnd != null ? cfg.businessEnd : 19,
      dupWindowDays: cfg.dupWindowDays != null ? cfg.dupWindowDays : 7,
      shortDescLen: cfg.shortDescLen != null ? cfg.shortDescLen : 15,
      sensitiveTypes: cfg.sensitiveTypes || { control: 1, bank: 1, equity: 1 },
      enableApproverRules: cfg.enableApproverRules !== false
    };
  }

  // --- small helpers ---------------------------------------------------------
  function pad(n) { return (n < 10 ? '0' : '') + n; }
  function fmtDate(d) { d = toDate(d); return d.getFullYear() + '-' + pad(d.getMonth() + 1) + '-' + pad(d.getDate()); }
  function fmtTs(d) { return fmtDate(d) + ' ' + pad(d.getHours()) + ':' + pad(d.getMinutes()); }
  function isHoliday(ts, holidays) {
    var key = fmtDate(ts);
    return (holidays || []).some(function (h) { return fmtDate(h) === key; });
  }

  return {
    SEVERITY_WEIGHT: SEVERITY_WEIGHT,
    normalizeAcctType: normalizeAcctType,
    ruleOverThreshold: ruleOverThreshold, ruleRoundThousand: ruleRoundThousand,
    ruleOffHours: ruleOffHours, ruleWeakDescription: ruleWeakDescription,
    ruleNoSupport: ruleNoSupport, ruleSensitiveAccount: ruleSensitiveAccount,
    ruleEntryPostGap: ruleEntryPostGap, ruleClosedPeriod: ruleClosedPeriod,
    ruleSodBreach: ruleSodBreach, ruleNearDuplicate: ruleNearDuplicate,
    RULES: RULES, evaluate: evaluate, score: score, assess: assess,
    assessRegister: assessRegister, contextFromConfig: contextFromConfig,
    fmtDate: fmtDate, fmtTs: fmtTs
  };
});

File: Week3/je_rules.py
"""
je_rules.py  -  THE DETERMINISTIC RULE ENGINE  (Week 3, Build 2)

A detective control over a manual-journal-entry register. Every flag here is
raised by a PURE, TESTED function - no AI touches this layer. The AI only comes
in afterwards (je_review.py) to judge the *ambiguous* flags and write a reviewer
note; a second agent then challenges it. The numbers, dates and booleans that
decide a flag are all computed here, so the control is versioned and auditable.

Right-size note: "was this posted on a Saturday?", "is the preparer also the
approver?", "does another entry have the same account+amount?" are all exact
tests. Sending them to an LLM would be slower, costlier and *non-deterministic* -
exactly what a control cannot be. Code flags; AI only explains the grey areas.

Schema of one journal entry (one row = one manual JE):
    je_id        str    document number
    entry_date   date   the effective/accounting date of the entry
    post_ts      datetime  when it was actually posted (date + time)
    period       str    accounting period, "YYYY-MM"
    account      str    primary GL account hit
    account_type str    e.g. "expense", "revenue", "control", "equity", "bank"
    amount       float  absolute magnitude of the entry (> 0)
    dr_cr        str    "DR" or "CR"
    preparer     str    who created it
    approver     str    who approved it ("" if unapproved)
    description  str    free-text memo
    has_support  bool   is a supporting document attached?
    dimensions   dict   {"department","class","location"} - "" if blank

Each rule returns a Flag or None. `evaluate()` runs them all for one entry;
`score()` turns the flags into a deterministic risk tier.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from datetime import date, datetime, timedelta
from typing import Callable, Optional


# ============================================================================
# Data model
# ============================================================================
@dataclass
class JournalEntry:
    je_id: str
    entry_date: date
    post_ts: datetime
    period: str
    account: str
    account_type: str
    amount: float
    dr_cr: str
    preparer: str
    approver: str
    description: str
    has_support: bool
    dimensions: dict = field(default_factory=dict)


@dataclass
class Flag:
    """One rule firing on one entry. `severity` weights the risk score."""
    rule: str
    severity: str          # "high" | "medium" | "low"
    detail: str            # human-readable, cites the exact facts that fired it


@dataclass
class RuleContext:
    """Everything a rule may need beyond the single entry it judges."""
    register: list          # all entries, for cross-entry rules (duplicates)
    closed_periods: frozenset = frozenset()
    holidays: frozenset = frozenset()          # frozenset[date]
    # thresholds (all finance-owned, versioned here rather than in a formula):
    approval_threshold: float = 50_000.0       # entries at/above need scrutiny
    round_min: float = 10_000.0                # ignore small round numbers
    gap_days: int = 30                         # entry-vs-post gap that looks stale
    business_start: int = 7                    # 07:00 local
    business_end: int = 19                     # 19:00 local
    dup_window_days: int = 7                   # near-duplicate look-back/ahead
    short_desc_len: int = 15                   # chars; below = uninformative
    # accounts that should almost never take a *manual* entry (posted by
    # subledgers / system processes). A manual JE here is the classic audit red
    # flag - see Week 8 (subledger-to-GL tie-out).
    sensitive_types: frozenset = frozenset({"control", "bank", "equity"})
    # run the approver-based rule (sod_breach). Set False for accounts with no JE
    # approval workflow: there the "no independent approver" branch has no signal
    # and would flag every entry. True is correct where journals are approved.
    enable_approver_rules: bool = True


SEVERITY_WEIGHT = {"high": 3, "medium": 2, "low": 1}


# ============================================================================
# The rules  -  each one pure: (JournalEntry, RuleContext) -> Flag | None
# ============================================================================
def rule_over_threshold(je: JournalEntry, ctx: RuleContext) -> Optional[Flag]:
    """Large manual entries carry the most misstatement risk; always reviewed."""
    if je.amount >= ctx.approval_threshold:
        return Flag("over_threshold", "medium",
                    f"amount {je.amount:,.0f} >= approval threshold "
                    f"{ctx.approval_threshold:,.0f}")
    return None


def rule_round_thousand(je: JournalEntry, ctx: RuleContext) -> Optional[Flag]:
    """
    Suspiciously round amounts (exact thousands) suggest an estimate or a plug
    rather than a figure traced to source. Small round numbers are normal, so we
    only flag round amounts at/above `round_min`.
    """
    if je.amount >= ctx.round_min and je.amount % 1000 == 0:
        return Flag("round_thousand", "low",
                    f"amount {je.amount:,.0f} is an exact multiple of 1,000")
    return None


def rule_off_hours(je: JournalEntry, ctx: RuleContext) -> Optional[Flag]:
    """
    Posted on a weekend, a holiday, or outside business hours. Legitimate at
    close, but a known pattern for entries made to avoid scrutiny.
    """
    ts = je.post_ts
    reasons = []
    if ts.weekday() >= 5:                       # 5=Sat, 6=Sun
        reasons.append("weekend")
    if ts.date() in ctx.holidays:
        reasons.append("holiday")
    if ts.hour < ctx.business_start or ts.hour >= ctx.business_end:
        reasons.append(f"{ts.hour:02d}:{ts.minute:02d} outside "
                       f"{ctx.business_start:02d}:00-{ctx.business_end:02d}:00")
    if reasons:
        return Flag("off_hours", "low",
                    "posted " + ", ".join(reasons) + f" ({ts:%Y-%m-%d %H:%M})")
    return None


def rule_weak_description(je: JournalEntry, ctx: RuleContext) -> Optional[Flag]:
    """
    Missing or uninformative memo. An entry a reviewer cannot understand from its
    description is not supportable. Generic fillers ("adjustment", "reclass",
    "to correct", "per <name>") are treated as no description.
    """
    desc = (je.description or "").strip()
    generic = {"adjustment", "adj", "reclass", "correction", "to correct",
               "true up", "true-up", "plug", "misc", "journal", "entry"}
    if not desc:
        return Flag("weak_description", "medium", "description is blank")
    low = desc.lower()
    if len(desc) < ctx.short_desc_len or low in generic:
        return Flag("weak_description", "low",
                    f"description uninformative: {desc!r}")
    return None


def rule_no_support(je: JournalEntry, ctx: RuleContext) -> Optional[Flag]:
    """
    No supporting document attached. Severity rises with size: an unsupported
    entry at/above the approval threshold is a material control gap.
    """
    if not je.has_support:
        sev = "high" if je.amount >= ctx.approval_threshold else "medium"
        return Flag("no_support", sev,
                    f"no supporting document (amount {je.amount:,.0f})")
    return None


def rule_sensitive_account(je: JournalEntry, ctx: RuleContext) -> Optional[Flag]:
    """
    Manual entry to an account normally maintained by a subledger or system
    process (AP/AR control, bank/clearing, equity/retained earnings). These are
    the reconciling items that scare auditors - a human moving a control account
    by hand outside the subledger.
    """
    if je.account_type in ctx.sensitive_types:
        return Flag("sensitive_account", "high",
                    f"manual JE to {je.account_type} account {je.account!r}")
    return None


def rule_entry_post_gap(je: JournalEntry, ctx: RuleContext) -> Optional[Flag]:
    """
    Large gap between the effective (entry) date and when it was actually posted.
    A stale entry posted long after its date can distort the wrong period and
    hints at cut-off manipulation.
    """
    gap = (je.post_ts.date() - je.entry_date).days
    if gap >= ctx.gap_days:
        return Flag("entry_post_gap", "medium",
                    f"{gap} days between entry date {je.entry_date} and posting "
                    f"{je.post_ts.date()}")
    return None


def rule_closed_period(je: JournalEntry, ctx: RuleContext) -> Optional[Flag]:
    """
    Back-dated into a period that is already closed. Posting to a closed period
    reopens signed-off numbers and is a serious control breach.
    """
    if je.period in ctx.closed_periods:
        return Flag("closed_period", "high",
                    f"posted into closed period {je.period}")
    return None


def rule_sod_breach(je: JournalEntry, ctx: RuleContext) -> Optional[Flag]:
    """
    Segregation-of-duties breach: the preparer is also the approver, or the entry
    was never independently approved. The whole point of maker-checker.
    """
    if je.approver and je.preparer == je.approver:
        return Flag("sod_breach", "high",
                    f"preparer and approver are the same person "
                    f"({je.preparer})")
    if not je.approver:
        return Flag("sod_breach", "medium",
                    f"no independent approver recorded (preparer {je.preparer})")
    return None


def rule_near_duplicate(je: JournalEntry, ctx: RuleContext) -> Optional[Flag]:
    """
    Another entry in the register hits the SAME account for the SAME amount
    within a short window - a possible double-post, or an amount split across
    entries to slide under an approval threshold.
    """
    for other in ctx.register:
        if other is je or other.je_id == je.je_id:
            continue
        if (other.account == je.account
                and abs(other.amount - je.amount) < 0.01
                and abs((other.post_ts.date() - je.post_ts.date()).days)
                <= ctx.dup_window_days):
            return Flag("near_duplicate", "medium",
                        f"matches {other.je_id}: same account {je.account!r}, "
                        f"amount {je.amount:,.0f}, within {ctx.dup_window_days}d")
    return None


# The registered rule set (order = the order flags are reported).
RULES: list[Callable[[JournalEntry, RuleContext], Optional[Flag]]] = [
    rule_over_threshold,
    rule_round_thousand,
    rule_off_hours,
    rule_weak_description,
    rule_no_support,
    rule_sensitive_account,
    rule_entry_post_gap,
    rule_closed_period,
    rule_sod_breach,
    rule_near_duplicate,
]


# ============================================================================
# Evaluation + deterministic scoring
# ============================================================================
# rules that depend on an approval workflow; skipped when enable_approver_rules is False.
_APPROVER_RULES = {rule_sod_breach}


def evaluate(je: JournalEntry, ctx: RuleContext) -> list[Flag]:
    """Run every rule against one entry; return the flags that fired, in order."""
    active = [r for r in RULES
              if ctx.enable_approver_rules or r not in _APPROVER_RULES]
    return [f for f in (rule(je, ctx) for rule in active) if f is not None]


def score(flags: list[Flag]) -> tuple[int, str]:
    """
    Deterministic risk score = sum of severity weights, mapped to a tier.
    The tier decides who the AI layer sends where:
        high   -> reviewer must look (any high-severity flag, or score >= 5)
        medium -> reviewer judges the residual risk (the AI's grey zone)
        low    -> logged, no review needed
        clear  -> nothing fired
    """
    if not flags:
        return 0, "clear"
    total = sum(SEVERITY_WEIGHT[f.severity] for f in flags)
    has_high = any(f.severity == "high" for f in flags)
    if has_high or total >= 5:
        return total, "high"
    if total >= 2:
        return total, "medium"
    return total, "low"


@dataclass
class Assessment:
    """The full deterministic verdict on one entry - the input to the AI layer."""
    je: JournalEntry
    flags: list[Flag]
    risk_score: int
    tier: str

    def flag_rules(self) -> list[str]:
        return [f.rule for f in self.flags]


def assess(je: JournalEntry, ctx: RuleContext) -> Assessment:
    flags = evaluate(je, ctx)
    s, tier = score(flags)
    return Assessment(je=je, flags=flags, risk_score=s, tier=tier)


def assess_register(register: list[JournalEntry], ctx: RuleContext) -> list[Assessment]:
    """
    Assess an entire register. We point the context at this register so the
    cross-entry rules (near-duplicate) can see every other line.
    """
    ctx.register = register
    return [assess(je, ctx) for je in register]
